jetty-http/src/main/java/org/eclipse/jetty/http/compression/HuffmanEncoder.java

```diff
@@ -0,0 +1,144 @@
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.http.compression;
+
+import java.nio.ByteBuffer;
+
+import static org.eclipse.jetty.http.compression.Huffman.CODES;
+import static org.eclipse.jetty.http.compression.Huffman.LCCODES;
+
+public class HuffmanEncoder
+{
+    private HuffmanEncoder()
+    {
+    }
+
+    public static int octetsNeeded(String s)
+    {
+        return octetsNeeded(CODES, s);
+    }
+
+    public static int octetsNeeded(byte[] b)
+    {
+        int needed = 0;
+        for (byte value : b)
+        {
+            int c = 0xFF & value;
+            needed += CODES[c][1];
+        }
+        return (needed + 7) / 8;
+    }
+
+    public static void encode(ByteBuffer buffer, String s)
+    {
+        encode(CODES, buffer, s);
+    }
+
+    public static void encode(ByteBuffer buffer, byte[] b)
+    {
+        encode(CODES, buffer, b);
+    }
+
+    public static int octetsNeededLowercase(String s)
+    {
+        return octetsNeeded(LCCODES, s);
+    }
+
+    public static void encodeLowercase(ByteBuffer buffer, String s)
+    {
+        encode(LCCODES, buffer, s);
+    }
+
+    private static int octetsNeeded(final int[][] table, String s)
+    {
+        int needed = 0;
+        int len = s.length();
+        for (int i = 0; i < len; i++)
+        {
+            char c = s.charAt(i);
+            if (Huffman.isIllegalCharacter(c))
+                return -1;
+            needed += table[c][1];
+        }
+
+        return (needed + 7) / 8;
+    }
+
+    /**
+     * @param table The table to encode by
+     * @param buffer The buffer to encode to
+     * @param s The string to encode
+     */
+    private static void encode(final int[][] table, ByteBuffer buffer, String s)
+    {
+        long current = 0;
+        int n = 0;
+        int len = s.length();
+        for (int i = 0; i < len; i++)
+        {
+            char c = s.charAt(i);
+            if (Huffman.isIllegalCharacter(c))
+                 throw new IllegalArgumentException();
+            int code = table[c][0];
+            int bits = table[c][1];
+
+            current <<= bits;
+            current |= code;
+            n += bits;
+
+            while (n >= 8)
+            {
+                n -= 8;
+                buffer.put((byte)(current >> n));
+            }
+        }
+
+        if (n > 0)
+        {
+            current <<= (8 - n);
+            current |= (0xFF >>> n);
+            buffer.put((byte)(current));
+        }
+    }
+
+    private static void encode(final int[][] table, ByteBuffer buffer, byte[] b)
```

Comment:
If our Huffman encode/decode are going to apply RFC9110 rules for what characters are acceptable, then do we need to encode bytes (or decode strings)?      Is it possible to have encode just do String->ByteBuffer and decode just do ByteBuffer->String ?

Reply:
Not sure what you mean here.

Reply:
I see two ways to make these APIs clean and the class behaviours regular.  Either:

a) The Huffman classes are just about encoding/decoding bytes and have no character rules.  The character rules are done at a higher level (either in hpack/qpack or a class in between).
OR b) The Huffman classes are specific to encoding/decoding HTTP headers (they are in http package after all) and thus they know about encoding/decoding characters specifically for HTTP headers.

We are option b) and I think we want to be as we can do more efficient char to byte conversions with the knowledge that we are ISO-8859-1 because we are only HTTP headers.  So given that these classes are only for http headers, then do we need methods to encode from bytes or decode to bytes?    

For example, is the method `public static void encode(ByteBuffer buffer, byte[] b)` ever actually used?  If so then why? by who?

If this class has rules about characters that can be encoded then it should only have the API: `public static void encode(ByteBuffer buffer, String s)` and not the byte version.